// @huggingface/inference/dist/commonjs/providers/featherless-ai.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatherlessAITextGenerationTask = exports.FeatherlessAIConversationalTask = void 0;
const InferenceOutputError_js_1 = require("../lib/InferenceOutputError.js");
const providerHelper_js_1 = require("./providerHelper.js");
const FEATHERLESS_API_BASE_URL = "https://api.featherless.ai";
class FeatherlessAIConversationalTask extends providerHelper_js_1.BaseConversationalTask {
    constructor() {
        super("featherless-ai", FEATHERLESS_API_BASE_URL);
    }
}
exports.FeatherlessAIConversationalTask = FeatherlessAIConversationalTask;
class FeatherlessAITextGenerationTask extends providerHelper_js_1.BaseTextGenerationTask {
    constructor() {
        super("featherless-ai", FEATHERLESS_API_BASE_URL);
    }
    preparePayload(params) {
        return {
            ...params.args,
            ...params.args.parameters,
            model: params.model,
            prompt: params.args.inputs,
        };
    }
    async getResponse(response) {
        if (typeof response === "object" &&
            "choices" in response &&
            Array.isArray(response?.choices) &&
            typeof response?.model === "string") {
            const completion = response.choices[0];
            return {
                generated_text: completion.text,
            };
        }
        throw new InferenceOutputError_js_1.InferenceOutputError("Expected Featherless AI text generation response format");
    }
}
exports.FeatherlessAITextGenerationTask = FeatherlessAITextGenerationTask;
