// @huggingface/inference/dist/esm/providers/featherless-ai.js
import { InferenceOutputError } from "../lib/InferenceOutputError.js";
import { BaseConversationalTask, BaseTextGenerationTask } from "./providerHelper.js";
const FEATHERLESS_API_BASE_URL = "https://api.featherless.ai";
export class FeatherlessAIConversationalTask extends BaseConversationalTask {
    constructor() {
        super("featherless-ai", FEATHERLESS_API_BASE_URL);
    }
}
export class FeatherlessAITextGenerationTask extends BaseTextGenerationTask {
    constructor() {
        super("featherless-ai", FEATHERLESS_API_BASE_URL);
    }
    preparePayload(params) {
        return {
            ...params.args,
            ...params.args.parameters,
            model: params.model,
            prompt: params.args.inputs,
        };
    }
    async getResponse(response) {
        if (typeof response === "object" &&
            "choices" in response &&
            Array.isArray(response?.choices) &&
            typeof response?.model === "string") {
            const completion = response.choices[0];
            return {
                generated_text: completion.text,
            };
        }
        throw new InferenceOutputError("Expected Featherless AI text generation response format");
    }
}
